wandb/sdk/launch/sweeps/scheduler_sweep.py

```diff
@@ -123,37 +115,48 @@ def _heartbeat(self, worker_id: int) -> None:
 
     def _run(self) -> None:
         # Go through all workers and heartbeat
-        for worker_id in self._workers.keys():
+        for worker_id in self._workers:
             self._heartbeat(worker_id)
-        try:
-            run: SweepRun = self._heartbeat_queue.get(
-                timeout=self._heartbeat_queue_timeout
-            )
-        except queue.Empty:
-            wandb.termlog(f"{LOG_PREFIX}No jobs in Sweeps RunQueue, waiting...")
-            time.sleep(self._heartbeat_queue_sleep)
-            return
-        # If run is already stopped just ignore the request
-        if run.state in [
-            SimpleRunState.DEAD,
-            SimpleRunState.UNKNOWN,
-        ]:
-            return
-        wandb.termlog(
-            f"{LOG_PREFIX}Converting Sweep Run (RunID:{run.id}) to Launch Job"
-        )
-        _ = self._add_to_launch_queue(
-            run_id=run.id,
-            entry_point=["python3", run.program] if run.program else None,
-            # Use legacy sweep utilities to extract args dict from agent heartbeat run.args
-            config={
-                "overrides": {
-                    "run_config": LegacySweepAgent._create_command_args(
-                        {"args": run.args}
-                    )["args_dict"]
-                }
-            },
-        )
+
+        for _worker_id in self._workers:
+            try:
+                run: SweepRun = self._heartbeat_queue.get(
+                    timeout=self._heartbeat_queue_timeout
+                )
+
+                # If run is already stopped just ignore the request
+                if run.state in [
+                    SimpleRunState.DEAD,
+                    SimpleRunState.UNKNOWN,
+                ]:
+                    if self.state == SchedulerState.STOPPED:
+                        # Scheduler hit the run_cap, let it launch runs then it will stop
+                        pass
+                    else:
+                        logging.debug(
+                            f"{LOG_PREFIX}Tring to launch [{run.id}] but it is already stopped"
```

Comment:
When can SimpleRunState be `UNKNOWN` is that always a stopped case?

Reply:
idk but our handling here has always been UKNOWN --> don't launch run

Reply:
Fair enough